### DISCUSSION TITLE
Is it possible to get an asynchronous method's return value without invoking GetAwaiter().GetResult() ?

### DISCUSSION THREAD

**Discussion by EtherealAO** [vfsfitvnm/frida-il2cpp-bridge#119] [Q&A]
I'm observing a game's communication by intercept
```
private UniTask<TResponse> CallInternal<TRequest, TResponse>(string apiName, AsyncUnaryCallDelegate<TRequest, TResponse> asyncUnaryCallDelegate, TRequest request, AsyncUnaryCallOption callOption, int hash, bool isSilent) { }
```
[UniTask\<T\>](https://github.com/Cysharp/UniTask) is a C# struct like ValueTask, can't use ``` new Il2Cpp.Object(args[2]).method("xxx") ```, and if UniTask<TResponse> awaits multiple times, it will throw an exception. So even I can invoke GetAwaiter().GetResult() to get TResponse, game will crashed because of unitask. Do anyone has advice?

**Comment by vfsfitvnm** [comment]
I don't know if I understood correctly, but you could just intercept `GetResult`? So you don't have to do it yourself

**Comment by EtherealAO** [comment]
Intercept UniTask.Awaiter.GetResult() failed, it never been invoked.
Intercept AsyncUnaryCall and try to invoke GetAwaiter().GetResult() failed, because awaiter is a struct, can't use ```<Il2Cpp.Object>awaiter.method("GetResult").invoke()```
Using ```Interceptor.attach(Il2Cpp.Image.corlib.class("System.Runtime.CompilerServices.TaskAwaiter`1").inflate(Il2Cpp.Image.corlib.class("System.Object")).method("GetResult"))``` failed, maybe because TaskAwaiter is a struct?
So I have to use Il2CppDumper to find out the offset and use ```Interceptor.attach(Il2Cpp.module.base.add("0x5569538"))``` instead. Fortunately this game's packet always ends with "Request" or "Response" and easy to filter.
Thank you for your advice

**Comment by vfsfitvnm** [comment]
(I'm moving this to discussions btw)

> Intercept AsyncUnaryCall and try to invoke GetAwaiter().GetResult() failed, because awaiter is a struct

You can call `box()` to get a `Il2Cpp.Object` from a `Il2Cpp.ValueType`

> Using Interceptor.attach(...).method("GetResult")) failed

Failed because you are intercepting a `Il2Cpp.Method`, you have to intercept its `virtualAddress` instead

**Comment by EtherealAO** [comment]
The struct I said is a C# struct, not NativeStruct or something.
> Failed because you are intercepting a Il2Cpp.Method, you have to intercept its virtualAddress instead

Of course I intercepted it's virtual address. I mean, if I intercept with ```Interceptor.attach(...).method("GetResult"))```, nothing will show in console and if i use ```Interceptor.attach(Il2Cpp.module.base.add("0x5569538")```, it works fine. So I think the problem might be with his type "struct " not "class"

**Reply by vfsfitvnm** [reply]
> The struct I said is a C# struct, not NativeStruct or something.

I don't get this? You said "awaiter is a struct", structs (`Il2Cpp.ValueType`s) do not have methods:
```ts
<Il2Cpp.Object>awaiter.method("GetResult").invoke() // won't work

awaiter.box().method("GetResult").invoke() // will work
```
> Of course I intercepted it's virtual address.

No, `Interceptor.attach([...].method("GetResult"))` will not intercept the method. but `Interceptor.attach([...].method("GetResult").virtualAddress)` will
